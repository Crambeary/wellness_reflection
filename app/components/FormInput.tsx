'use client';

import React, { useEffect, useRef } from 'react';

interface FormInputProps {
  label: string;
  type?: string;
  id: string;
  value: string;
  fieldType?: string;
  onChange: (event: React.ChangeEvent<HTMLInputElement | HTMLDivElement> | React.FormEvent<HTMLDivElement>) => void;
  disabled?: boolean;
}

const FormInput: React.FC<FormInputProps> = ({ 
  label, 
  type = "text", 
  id, 
  value, 
  onChange, 
  fieldType = "input",
  disabled = false
}) => {
  const contentRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (contentRef.current && fieldType === "textarea") {
      contentRef.current.textContent = value;
    }
  }, [value, fieldType]);

  return (
    <>
      {fieldType === "input" ? (
        <div className="input-group mb-3">
          <label htmlFor={id} className="input-group-text">{label}</label>
          <input
            className="form-control"
            type={type}
            id={id}
            value={value}
            onChange={onChange}
            spellCheck={false}
            disabled={disabled}
          />
          {type === 'time' && (
            <button className='btn btn-outline-danger' onClick={() => {
              console.log('clearing');
              onChange({ target: { id, value: '' } } as React.ChangeEvent<HTMLInputElement>);
              }}
              disabled={disabled}
              >
              X
            </button>
          )}
        </div>
      ) : (
        <div className="mb-3">
          <label htmlFor={id} className="form-label">{label}</label>
          <div 
            ref={contentRef}
            contentEditable={!disabled}
            suppressContentEditableWarning={true}
            className="form-control"
            id={id}
            onBlur={(e) => {
              onChange(e);
            }}
            spellCheck={false}
          />
        </div>
      )}
    </>
  );
};

export default FormInput;
