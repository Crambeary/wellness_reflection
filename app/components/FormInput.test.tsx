import renderer from "react-test-renderer";
import FormInput from "./FormInput";

describe('FormInput', () => {
  it("renders correctly", () => {
    const mockOnChange = jest.fn();
    const component = renderer.create(
      <FormInput
        label="Test Label"
        id="test-id"
        value="Test Value"
        onChange={mockOnChange}
        fieldType="input"
        disabled={false}
      />,
    );
    
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
