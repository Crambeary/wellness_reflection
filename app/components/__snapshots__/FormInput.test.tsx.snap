// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput renders correctly 1`] = `
<div
  className="input-group mb-3"
>
  <label
    className="input-group-text"
    htmlFor="test-id"
  >
    Test Label
  </label>
  <input
    className="form-control"
    disabled={false}
    id="test-id"
    onChange={[MockFunction]}
    spellCheck={false}
    type="text"
    value="Test Value"
  />
</div>
`;
